aws/rust-runtime/aws-sigv4/src/http_request/error.rs

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use http::header::{InvalidHeaderName, InvalidHeaderValue};
+use std::error::Error;
+use std::fmt;
+use std::str::Utf8Error;
+
+#[derive(Debug)]
+enum SigningErrorKind {
+    FailedToCreateCanonicalRequest { source: CanonicalRequestError },
+}
+
+/// Error signing request
+#[derive(Debug)]
+pub struct SigningError {
+    kind: SigningErrorKind,
+}
+
+impl fmt::Display for SigningError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self.kind {
+            SigningErrorKind::FailedToCreateCanonicalRequest { .. } => {
+                write!(f, "failed to create canonical request")
+            }
+        }
+    }
+}
+
+impl Error for SigningError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match &self.kind {
+            SigningErrorKind::FailedToCreateCanonicalRequest { source } => Some(source),
+        }
+    }
+}
```

Comment:
I believe this follows https://github.com/awslabs/smithy-rs/issues/1345?

Reply:
Yes. The source doesn't get written in the `Display` impl so that it's not duplicated for error reporters that follow the source chain.